opm/simulators/flow/ReservoirCouplingMpiTraits.hpp

```diff
@@ -0,0 +1,154 @@
+        //   a static flag.
+        std::call_once(flag_, [] {
+
+            constexpr std::size_t N = sizeof...(Members);
+
+            // Array of block lengths (number of elements per field)
+            //  NOTE: The block length is set to 1 for each field (blk.fill(1)),
+            //        but is strictly not necessary currently, as the processMember()
+            //        function will always set the block length anyway.
+            //        However, it is kept here for clarity and future-proofing.
+            std::array<int,      N> blk{};           blk.fill(1);
+            // Array of displacements for each field
+            std::array<MPI_Aint, N> disp{};
+            // Array of MPI data types for each field
+            std::array<MPI_Datatype, N> types{};
+
+            // Create a dummy instance of the struct to get the base address
+            Struct dummy{};
+            // Get the base address of the dummy instance
+            MPI_Aint base{};
+            MPI_Get_address(&dummy, &base);
+
+            std::size_t i = 0;
+            // fold expression over the member pointer pack
+            ( processMember<Members>(dummy, base, disp, blk, types, i++), ... );
+
+            MPI_Datatype tmp;
+            // Create the MPI datatype
+            MPI_Type_create_struct(N, blk.data(), disp.data(), types.data(), &tmp);
+            // Resize the datatype to account for possible padding issues
+            MPI_Type_create_resized(tmp, 0, sizeof(Struct), &type_);
+            MPI_Type_commit(&type_);
+            // Free the temporary datatype
+            // Note: This is necessary to avoid memory leaks, as the temporary datatype
+            //       created by MPI_Type_create_struct is not automatically freed.
+            MPI_Type_free(&tmp);
+        });
+        return type_;
+    }
+
+    static constexpr bool is_intrinsic = false;
+
+  private:
+    // --- helper to recognise std::array ----------------------------------------
+    template<class T> struct is_std_array : std::false_type {};
+    template<class T, std::size_t N>
+    struct is_std_array<std::array<T, N>> : std::true_type {};
+
+    template<auto Member, class Dummy>
+    static void processMember(Dummy& d, MPI_Aint base,
+                              std::array<MPI_Aint,     sizeof...(Members)>& disp,
+                              std::array<int,          sizeof...(Members)>& blk,
+                              std::array<MPI_Datatype, sizeof...(Members)>& types,
+                              std::size_t idx)
+    {
+        using MemberT = std::remove_reference_t<decltype(d.*Member)>;
+
+        MPI_Get_address(&(d.*Member), &disp[idx]);
+        disp[idx] -= base;
+
+        if constexpr (std::is_array_v<MemberT>) {
+            // C array  T[N]
+            using Elem = std::remove_extent_t<MemberT>;
+            blk  [idx] = std::extent_v<MemberT>;
+            types[idx] = MPITraits<Elem>::getType();
+        }
```

Comment:
This should also work if T is not an array and give the same results as the else branch (you still need to handle num). Can we simplify this?

What branches are actually used? This looks quite general, but the surrouding class is not.